Added "Porting PyTorch model to JAX" tutorial (#71)
jax-ml/jax-ai-stack: docs/JAX_porting_PyTorch_model.md

- blocks: list of MaxViT blocks
- classifier: adaptive average pooling, few linear layers and Tanh activation.

```{code-cell} ipython3
from torchvision.models import maxvit_t, MaxVit_T_Weights

torch_model = maxvit_t(weights=MaxVit_T_Weights.IMAGENET1K_V1)
```

We can use `flax.nnx.display` to display the model's architecture:
```python
nnx.display(torch_model)
```

+++

We can see that there are four MaxViT blocks in the model and each block contains:
- MaxViT layers: two layers for blocks 0, 1, 3 and five layers for the block 4

```{code-cell} ipython3
len(torch_model.blocks), [len(b.layers) for b in torch_model.blocks]
```

A MaxViT layer is composed of: MBConv, window_attention as `PartitionAttentionLayer` and grid_attention as `PartitionAttentionLayer`.

```{code-cell} ipython3
[[mod.__class__.__name__ for mod in maxvit_layer.layers] for b in torch_model.blocks for maxvit_layer in b.layers]
```

### Inference on data

Let's check the model on dummy input and on a real image

```{code-cell} ipython3
import torch

torch_model.eval()
with torch.inference_mode():
    x = torch.rand(2, 3, 224, 224)
    output = torch_model(x)

> print(output.shape)

Author: It shows the shape of logits: e.g (2, 1000)

wget "https://raw.githubusercontent.com/pytorch/vision/refs/heads/main/gallery/assets/imagenet_class_index.json" -O imagenet_class_index.json
```

```{code-cell} ipython3
import matplotlib.pyplot as plt

import json
from torchvision.io import read_image

preprocess = MaxVit_T_Weights.IMAGENET1K_V1.transforms()

with open("imagenet_class_index.json") as labels_file:
  labels = json.load(labels_file)

dog1 = read_image("dog1.jpg")
tensor = preprocess(dog1)

torch_model.eval()
with torch.inference_mode():
  output = torch_model(tensor.unsqueeze(dim=0))

class_id = output.argmax(dim=1).item()

print(f"Prediction for the Dog: {labels[str(class_id)]}, score: {output.softmax(dim=-1)[0, class_id]}")

plt.title(f"{labels[str(class_id)]}\nScore: {output.softmax(dim=-1)[0, class_id]}")
plt.imshow(dog1.permute(1, 2, 0))
```

## Port MaxViT model to JAX

Let us now define all the necessary model parts in JAX using Flax module. PyTorch model's implementation can be found [here](https://pytorch.org/vision/stable/_modules/torchvision/models/maxvit.html#maxvit_t), we can now list the modules and helper functions we implement in JAX:

> - Conv2dNormActivation
> - SqueezeExcitation
> - StochasticDepth
> - MBConv
> - RelativePositionalMultiHeadAttention
> - SwapAxes
> - WindowPartition
> - WindowDepartition
> - PartitionAttentionLayer
> - MaxVitLayer
> - MaxVitBlock
> - MaxViT

Reviewer: ```suggestion
- `Conv2dNormActivation`
- `SqueezeExcitation`
- `StochasticDepth`
- `MBConv`
- `RelativePositionalMultiHeadAttention`
- `SwapAxes`
- `WindowPartition`
- `WindowDepartition`
- `PartitionAttentionLayer`
- `MaxVitLayer`
- `MaxVitBlock`
- `MaxViT`
```

For consistency.

I also wonder if listing these items here with no context here is helpful, but I think the alternative (giving a short description of each item) might be too much information at this point, seeing as you will detail the implementations below.

Author: I can rearrange them in the following way:
```
To port the [PyTorch implementation of the MaxVit model](https://github.com/pytorch/vision/blob/945bdad7523806b15d3740ce6ace2fced9ef9d3b/torchvision/models/maxvit.py#L568) in JAX using the Flax module, we will implement the following required modules:

- `MaxViT`
  - `MaxVitBlock`
    - `MaxVitLayer`
      - `MBConv`
        - `Conv2dNormActivation`
        - `SqueezeExcitation`
      - `PartitionAttentionLayer`
        - `RelativePositionalMultiHeadAttention`
        - `WindowDepartition`
        - `WindowPartition`
        - `SwapAxes`
        - `StochasticDepth`
```

Reviewer: Ah that does help!

> Flax NNX module is very similar to PyTorch `torch.nn` module and we can simply map the following modules between PyTorch and Flax:

Reviewer: ```suggestion
The Flax NNX module is very similar to PyTorch `torch.nn` module and we can map the following modules between PyTorch and Flax:
```

- `nn.Sequential` and `nn.ModuleList` -> `nnx.Sequential`
- `nn.Linear` -> `nnx.Linear`
- `nn.Conv2d` -> `nnx.Conv`
- `nn.BatchNorm2d` -> `nnx.BatchNorm`

> - Activations like `nn.ReLU` -> `lambda x: nnx.relu(x)`

Reviewer 2: No need to wrap in a `lambda` here:
```suggestion
- Activations like `nn.ReLU` -> `nnx.relu`
```

> - Pooling layers like `nn.AvgPool2d` -> `lambda x: nnx.avg_pool(x, ...)`

Reviewer 2: Maybe the same here? I'm not sure what `...` means.

Reviewer 2: Then maybe something like this?
```suggestion
- Pooling layers like `nn.AvgPool2d(...)` -> `lambda x: nnx.avg_pool(x, ...)`
- ```

- `nn.AdaptiveAvgPool2d(1)` -> `lambda x: nnx.avg_pool(x, (x.shape[1], x.shape[2]))`, x is in NHWC format

> - `nn.Flatten()` -> `lambda x: x.reshape(x.shape[0], -1)`

Reviewer 2: maybe use `jax.vmap(jnp.ravel)` here? Or is that too magic?

Author: yes, this is too magic.

> If PyTorch model defines a learnable parameter and a buffer:

Reviewer: ```suggestion
If the PyTorch model defines a learnable parameter and a buffer:
```

```python
class Model(nn.Module):
    def __init__(self, ...):
        ...
        self.p = nn.Parameter(torch.ones(10))
        self.register_buffer("b", torch.ones(5))
```

> equivalent code in Flax would be

Reviewer: ```suggestion
the equivalent code in Flax would be
```

```python
class Buffer(nnx.Variable):
    pass

class Model(nnx.Module):
    def __init__(self, ...):
        ...
        self.p = nnx.Param(jnp.ones((10,)))
        self.b = Buffer(jnp.ones(5))
```

To inspect NNX module's learnable parameters and buffers, we can use `nnx.state`:
```python
nnx_module = ...
for k, v in nnx.state(nnx_module, nnx.Param).flat_state().items():
    print(
        k,
        v.value.mean() if v.value is not None else None
    )

for k, v in nnx.state(nnx_module, (nnx.BatchStat, Buffer)).flat_state().items():
    print(
        k,
        v.value.mean() if v.value.dtype == "float32" else v.value.sum()
    )
```

> Equivalent PyTorch code is:

Reviewer: ```suggestion
The equivalent PyTorch code is:
```

```python
torch_module = ...

for m, p in torch_module.named_parameters():
    print(m, p.detach().mean())

for m, b in torch_module.named_buffers():
    print(
        m,
        b.mean() if b.dtype == torch.float32 else b.sum()
    )
```

+++

> Please note the some tiny differences between `torch.nn` and Flax when porting models:
> - We should pass `rngs` to all NNX modules with parameters: e.g. `nnx.Linear(..., rngs=nnx.Rngs(0))`
> - Convolution 2D in Flax:
>   - we need explicitly define `kernel_size`, `strides` as two ints tuples, e.g. `(3, 3)`
>   - if `padding` should be defined as integer, e.g. 2, it should be defined as `((2, 2), (2, 2))`
> - Batch Normalization: momentum in `torch.nn` is defined as `1.0 - momentum` in Flax
> - 4D input array in Flax should be in NHWC format, i.e. of shape (N, H, W, C) compared to NCHW format (or (N, C, H, W) shape) in PyTorch.

Reviewer: ```suggestion
Please note there are some differences between `torch.nn` and Flax when porting models:
- We should pass `rngs` to all NNX modules with parameters: e.g. `nnx.Linear(..., rngs=nnx.Rngs(0))`
- For a 2d convolution:
  - In Flax, we need to explicitly define `kernel_size`, `strides` as two ints tuples, e.g. `(3, 3)`
  - If `padding` should be defined as integer in PyTorch, e.g. 2, it should be defined as `((2, 2), (2, 2))` in Flax.
- For a batch normalization: the momentum in `torch.nn` is defined as `1.0 - momentum` in Flax
- 4D input arrays in Flax should be in NHWC format, i.e. of shape (N, H, W, C) compared to NCHW format (or (N, C, H, W) shape) in PyTorch.
```

Please make sure these are correct especially the padding item, the phrasing is what I am focusing on.

+++

Below we implement one by one all the modules from the above list and add a simple forward pass check.

> 
> ```{code-cell} ipython3
> # Let's first implement equivalent of nn.Identity
> class Identity(nnx.Module):
>     def __call__(self, x: jax.Array) -> jax.Array:
>         return x
> ```

Reviewer: ````suggestion
Let's first implement equivalent of `nn.Identity`:

```{code-cell} ipython3
class Identity(nnx.Module):
    def __call__(self, x: jax.Array) -> jax.Array:
        return x
```
````

    relative_coords = relative_coords * jnp.array((2 * width - 1, 1))

    return relative_coords.sum(-1)

output = _get_relative_position_index(13, 12)
expected = pytorch_get_relative_position_index(13, 12)
assert (output == jnp.asarray(expected)).all()
```

```{code-cell} ipython3
import math

class Buffer(nnx.Variable):
    pass

class RelativePositionalMultiHeadAttention(nnx.Module):
    def __init__(
        self,
        feat_dim: int,
        head_dim: int,
        max_seq_len: int,
        rngs: nnx.Rngs = nnx.Rngs(0),
    ):
        if feat_dim % head_dim != 0:
            raise ValueError(f"feat_dim: {feat_dim} must be divisible by head_dim: {head_dim}")

        self.n_heads = feat_dim // head_dim
        self.head_dim = head_dim
        self.size = int(math.sqrt(max_seq_len))
        self.max_seq_len = max_seq_len

        self.to_qkv = nnx.Linear(feat_dim, self.n_heads * self.head_dim * 3, rngs=rngs)
        self.scale_factor = feat_dim**-0.5

        self.merge = nnx.Linear(self.head_dim * self.n_heads, feat_dim, rngs=rngs)

        self.relative_position_index = Buffer(_get_relative_position_index(self.size, self.size))

>         # initialize with truncated normal the bias

Reviewer: ```suggestion
        # initialize with truncated normal bias
```

y = mod(x)
y.shape
```

#### `SwapAxes`, `WindowPartition`, `WindowDepartition` implementations

PyTorch source implementation: https://pytorch.org/vision/stable/_modules/torchvision/models/maxvit.html

```{code-cell} ipython3
class SwapAxes(nnx.Module):
    def __init__(self, a: int, b: int):
        self.a = a
        self.b = b

    def __call__(self, x: jax.Array) -> jax.Array:
        res = jnp.swapaxes(x, self.a, self.b)
        return res

class WindowPartition(nnx.Module):
    def __call__(self, x: jax.Array, p: int) -> jax.Array:
        # Output array with expected layout of [B, H/P, W/P, P*P, C].
        B, H, W, C = x.shape
        P = p
        # chunk up H and W dimensions
        x = x.reshape((B, H // P, P, W // P, P, C))
        x = jnp.permute_dims(x, (0, 1, 3, 2, 4, 5))
        # colapse P * P dimension
        x = x.reshape((B, (H // P) * (W // P), P * P, C))
        return x

class WindowDepartition(nnx.Module):
    def __call__(self, x: jax.Array, p: int, h_partitions: int, w_partitions: int) -> jax.Array:
        # Output array with expected layout of [B, H, W, C].
        B, G, PP, C = x.shape
        P = p
        HP, WP = h_partitions, w_partitions

>         # split P * P dimension into 2 P tile dimensionsa

Reviewer: ```suggestion
        # split P * P dimension into 2 P tile dimensions
```

        x = x.reshape((B, HP, WP, P, P, C))
        # permute into B, HP, P, WP, P, C
        x = jnp.permute_dims(x, (0, 1, 3, 2, 4, 5))
        # reshape into B, H, W, C
        x = x.reshape((B, HP * P, WP * P, C))
        return x
```

```{code-cell} ipython3
x = jnp.ones((3, 4, 5, 6))
mod = SwapAxes(1, 2)
y = mod(x)
assert y.shape == (3, 5, 4, 6)

x = jnp.ones((4, 128, 128, 3))
mod = WindowPartition()
y = mod(x, p=16)
assert y.shape == (4, (128 // 16) * (128 // 16), 16 * 16, 3)

x = jnp.ones((4, (128 // 16) * (128 // 16), 16 * 16, 3))
mod = WindowDepartition()
y = mod(x, p=16, h_partitions=128 // 16, w_partitions=128 // 16)
assert y.shape == (4, 128, 128, 3)
```

#### `PartitionAttentionLayer` implementation

PyTorch source implementation: https://pytorch.org/vision/stable/_modules/torchvision/models/maxvit.html

```{code-cell} ipython3

> # Layer for partitioning the input tensor into non-overlapping windows and applying attention to each window.
> 
> class PartitionAttentionLayer(nnx.Module):
>     def __init__(
>         self,
>         in_channels: int,
>         head_dim: int,
>         # partitioning parameters
>         partition_size: int,
>         partition_type: str,
>         # grid size needs to be known at initialization time
>         # because we need to know hamy relative offsets there are in the grid
>         grid_size: Tuple[int, int],
>         mlp_ratio: int,
>         activation_layer: Callable,
>         norm_layer: Callable[..., nnx.Module],
>         attention_dropout: float,
>         mlp_dropout: float,
>         p_stochastic_dropout: float,
>         rngs: nnx.Rngs = nnx.Rngs(0),
>     ):

Reviewer: ```suggestion
class PartitionAttentionLayer(nnx.Module):
    def __init__(
        self,
        in_channels: int,
        head_dim: int,
        # partitioning parameters
        partition_size: int,
        partition_type: str,
        # grid size needs to be known at initialization time
        # because we need to know hamy relative offsets there are in the grid
        grid_size: Tuple[int, int],
        mlp_ratio: int,
        activation_layer: Callable,
        norm_layer: Callable[..., nnx.Module],
        attention_dropout: float,
        mlp_dropout: float,
        p_stochastic_dropout: float,
        rngs: nnx.Rngs = nnx.Rngs(0),
    ):
    """Layer for partitioning the input tensor into non-overlapping windows and applying attention to each window.
    """
```

                    expansion_ratio=expansion_ratio,
                    stride=stride,
                    norm_layer=norm_layer,
                    activation_layer=activation_layer,
                    head_dim=head_dim,
                    mlp_ratio=mlp_ratio,
                    mlp_dropout=mlp_dropout,
                    attention_dropout=attention_dropout,
                    partition_size=partition_size,
                    grid_size=self.grid_size,
                    p_stochastic_dropout=p,
                ),
            )
        self.layers = nnx.Sequential(*layers)

    def __call__(self, x: jax.Array) -> jax.Array:
        return self.layers(x)
```

```{code-cell} ipython3
x = jnp.ones((4, 224, 224, 3))

input_grid_size = (224, 224)
norm_layer = partial(nnx.BatchNorm, epsilon=1e-3, momentum=0.99)

mod = MaxVitBlock(
    3, 36, squeeze_ratio=0.25, expansion_ratio=4,
    norm_layer=norm_layer, activation_layer=nnx.gelu,
    head_dim=6, mlp_ratio=4, mlp_dropout=0.5, attention_dropout=0.4,
    partition_size=7, input_grid_size=input_grid_size,
    n_layers=2,
    p_stochastic=[0.0, 0.2],
)

y = mod(x)
print(y.shape)
```

#### `MaxVit` implementation

> Finally, we can assemble evreything together and define MaxVit model.

Reviewer: ```suggestion
Finally, we can assemble everything together and define the MaxVit model.
```

                    f"Consider changing the partition size or the input size.\n"
                    f"Current configuration yields the following block input sizes: {block_input_sizes}."
                )

        # stem
        self.stem = nnx.Sequential(
            Conv2dNormActivation(
                input_channels,
                stem_channels,
                3,
                stride=2,
                norm_layer=norm_layer,
                activation_layer=activation_layer,
                bias=False,
                rngs=rngs,
            ),
            Conv2dNormActivation(
                stem_channels,
                stem_channels,
                3,
                stride=1,
                norm_layer=None,
                activation_layer=None,
                bias=True,
                rngs=rngs,
            ),
        )

        # account for stem stride
        input_size = _get_conv_output_shape(input_size, kernel_size=3, stride=2, padding=1)
        self.partition_size = partition_size

        # blocks
        blocks = []
        in_channels = [stem_channels] + block_channels[:-1]
        out_channels = block_channels

>         # precompute the stochastich depth probabilities from 0 to stochastic_depth_prob

Reviewer: ```suggestion
        # precompute the stochastic depth probabilities from 0 to stochastic_depth_prob
```

def test_modules(
    nnx_module, torch_module, torch_input, atol=1e-3, mode="eval", permute_torch_input=True, device="cuda"
):
    assert torch_input.ndim == 4
    assert mode in ("eval", "train")

    torch_input = torch_input.to(device)
    torch_module = torch_module.to(device)

    if mode == "eval":
        torch_module.eval()
        nnx_module.eval()
    else:
        torch_module.train()
        nnx_module.train()

    with torch.inference_mode(mode=mode=="eval"):
        torch_output = torch_module(torch_input)

    if permute_torch_input:
        torch_input = torch_input.permute(0, 2, 3, 1)

    jax_input = jnp.asarray(torch_input, device=jax.devices(device)[0])
    jax_output = nnx_module(jax_input)
    assert jax_output.device == jax.devices(device)[0]

    torch_output = torch_output.detach()
    if permute_torch_input and torch_output.ndim == 4:
        torch_output = torch_output.permute(0, 2, 3, 1)
    jax_expected = jnp.asarray(torch_output)

    assert jnp.allclose(jax_output, jax_expected, atol=atol), (
        jnp.abs(jax_output - jax_expected).max(),
        jnp.abs(jax_output - jax_expected).mean(),
    )

t2f = Torch2Flax()
```

> Let us now test our JAX modules. We only test the result of the forward pass in the inference mode such that we avoid disrcrepancies related to random layers like Dropout, StochasticDepth, etc.
> By default, we use absolute error tolerence `1e-3` when comparing the JAX output vs expected PyTorch result.
> For larger modules we set the device to CPU for JAX model to execute on in order to reduce the errors between CPU and CUDA.

Reviewer: (I think mode is meant to be model, but maybe I am wrong?)
```suggestion
Let us now test our JAX modules. We only test the result of the forward pass in the inference model such that we avoid discrepancies related to random layers like `Dropout`, `StochasticDepth`, etc.
By default, we use absolute error tolerence `1e-3` when comparing the JAX output vs expected PyTorch result.
For larger modules we set the device to CPU for the JAX model to execute on in order to reduce the errors between CPU and CUDA.
```

Author: No, inference mode it is a term meaning that we do not track grads on the forward pass and do not update buffers.

])

test_modules(nnx_module, torch_module, torch.randn(4, 36, 224, 224), device="cpu")
```

```{code-cell} ipython3
from torchvision.models.maxvit import MaxVitBlock as PyTorchMaxVitBlock

norm_layer = partial(nn.BatchNorm2d, eps=1e-3, momentum=0.01)
torch_module = PyTorchMaxVitBlock(
    64, 128, squeeze_ratio=0.25, expansion_ratio=4,
    norm_layer=norm_layer, activation_layer=nn.GELU,
    head_dim=32, mlp_ratio=4, mlp_dropout=0.0, attention_dropout=0.0,
    partition_size=7, input_grid_size=(56, 56),
    n_layers=2,
    p_stochastic=[0.13333333333333333, 0.2],
)

norm_layer = partial(nnx.BatchNorm, epsilon=1e-3, momentum=0.99)
nnx_module = MaxVitBlock(
    64, 128, squeeze_ratio=0.25, expansion_ratio=4,
    norm_layer=norm_layer, activation_layer=nnx.gelu,
    head_dim=32, mlp_ratio=4, mlp_dropout=0.0, attention_dropout=0.0,
    partition_size=7, input_grid_size=(56, 56),
    n_layers=2,
    p_stochastic=[0.13333333333333333, 0.2],
)

t2f.copy_module(torch_module, nnx_module, skip_modules=[
    "WindowPartition", "WindowDepartition", "SwapAxes", "StochasticDepth",
])

test_modules(nnx_module, torch_module, torch.randn(4, 64, 56, 56), device="cpu")
```

> Finally, we can check MaxVit implementation. Note that we raised absolute tolerence to `1e-1` when compare JAX output logits vs PyTorch expected logits.

Reviewer: ```suggestion
Finally, we can check the MaxVit implementation. Note that we raised the absolute tolerence to `1e-1` when comparing JAX output logits vs PyTorch expected logits.
```

from torchvision.io import read_image

preprocess = MaxVit_T_Weights.IMAGENET1K_V1.transforms()

with open("imagenet_class_index.json") as labels_file:
  labels = json.load(labels_file)

dog1 = read_image("dog1.jpg")
tensor = preprocess(dog1).unsqueeze(dim=0)

torch_model.eval()
with torch.inference_mode():
    torch_output = torch_model(tensor)

torch_class_id = torch_output.argmax(dim=1).item()

jax_array = jnp.asarray(tensor.permute(0, 2, 3, 1), device=jax.devices("cpu")[0])
flax_model.eval()
flax_output = flax_model(jax_array)

flax_class_id = torch_output.argmax(axis=1).item()

print("Prediction for the Dog:")
print(f"- PyTorch model result: {labels[str(torch_class_id)]}, score: {torch_output.softmax(axis=1)[0, torch_class_id]}")
print(f"- Flax model result: {labels[str(flax_class_id)]}, score: {jax.nn.softmax(flax_output, axis=1)[0, flax_class_id]}")

plt.subplot(121)
plt.title(f"{labels[str(torch_class_id)]}\nScore: {torch_output.softmax(dim=-1)[0, class_id]:.4f}")
plt.imshow(dog1.permute(1, 2, 0))

plt.subplot(122)
plt.title(f"{labels[str(flax_class_id)]}\nScore: {jax.nn.softmax(flax_output, axis=1)[0, flax_class_id]:.4f}")
plt.imshow(dog1.permute(1, 2, 0))
```

> 
> ```{code-cell} ipython3
> # Let's compute cosine distance between the logits:

Reviewer: ```suggestion

Let's compute cosine distance between the logits:

```{code-cell} ipython3
```
